include/exiv2/slice.hpp

```diff
@@ -124,361 +79,44 @@ namespace Exiv2
-                const size_t new_begin = begin + this->begin_;
-                const size_t new_end = this->begin_ + end;
-                if (new_end > this->end_) {
-                    throw std::out_of_range("Invalid input parameters to slice");
-                }
-                return slice_type(storage_.data_, new_begin, new_end);
-            }
-
-        protected:
-            /*!
-             * Stores a reference to the actual data.
-             */
-            storage_type<data_type> storage_;
-        };
-
-        /*!
-         * This class provides all public-facing non-const-qualified methods of
-         * slices. It only re-implements the const-qualified versions as
-         * non-const.
-         */
-        template <template <typename> class storage_type, typename data_type>
-        struct MutableSliceBase : public ConstSliceBase<storage_type, data_type>
-        {
-            typedef typename ConstSliceBase<storage_type, data_type>::iterator iterator;
-            typedef typename ConstSliceBase<storage_type, data_type>::const_iterator const_iterator;
-            typedef typename ConstSliceBase<storage_type, data_type>::value_type value_type;
-
-            /*!
-             * Forwards everything to the constructor of const_slice_base
-             *
-             * @todo use using once we have C++11
-             */
-            MutableSliceBase(data_type& data, size_t begin, size_t end)
-                : ConstSliceBase<storage_type, data_type>(data, begin, end)
-            {
-            }
-
-            /*!
-             * Obtain a reference to the element with the specified index in the
-             * slice.
+             * @param[in,out] begin, end new bounds of a subSlice, this function
+             *     adds the current `begin_` and `end_` to these safely.
              *
-             * @throw std::out_of_range when index is out of bounds of the slice
-             */
-            value_type& at(size_t index)
-            {
-                this->rangeCheck(index);
-                return this->storage_.unsafeAt(this->begin_ + index);
-            }
-
-            const value_type& at(size_t index) const
-            {
-                // TODO: use using base_type::at once we have C++11
-                return base_type::at(index);
-            }
-
-            /*!
-             * Obtain an iterator to the first element in the slice.
-             */
-            iterator begin() noexcept
-            {
-                return this->storage_.unsafeGetIteratorAt(this->begin_);
-            }
-
-            /*!
-             * Obtain an iterator to the first element beyond the slice.
+             * @throw std::out_of_range when `begin` or `end` will access an
+             *     element outside of the slice
              */
-            iterator end() noexcept
+            inline void calculateSubSliceBounds(size_t& begin, size_t& end) const
```

Comment:
If I am not mistaken we could safely remove all the `inline` specifiers from this file. According to cppreference:

```
A function defined entirely inside a class/struct/union definition, whether it's a member function or a non-member friend function, is implicitly an inline function.
```

Reply:
What?! C++ never ceases to amaze with occasional "interesting" decisions.

Let's see if the compilers agree ;-)

Reply:
Hehe, I think this particular decision is there even before C++11. For years I have been making comments in PRs about this particular point 😉

Reply:
A lot of the "interesting" decisions predate C++11, like this one:
``` C++
char c = 0;
const int& i = c;
c = 5;
assert(i == 0);
```
(No you don't get an assertion failure...)